File: implementations/java-springboot/examples/src/main/java/com/example/accountservice/account/application/query/GetSpendingAnalysisService.java
package com.example.accountservice.account.application.query;

import com.example.accountservice.account.domain.AccountException;
import com.example.accountservice.account.domain.AccountFindQuery;
import com.example.accountservice.account.domain.SpendingAnalysis;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Account (unlike Payment/Refund) has an ownerId column directly on it, so ownership is verified
 * with a single {@code AccountQuery.findAccounts} lookup — simpler than Payment/Refund's two-hop
 * verification. Mirrors {@code GetAccountService}'s structure.
 */
@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class GetSpendingAnalysisService {

    private final AccountQuery accountQuery;
    private final SpendingAnalysisQuery spendingAnalysisQuery;

    public SpendingAnalysisResult getSpendingAnalysis(
            String accountId, String requesterId, String analysisMonth) {
        accountQuery
                .findAccounts(new AccountFindQuery(0, 1, accountId, requesterId, null))
                .accounts()
                .stream()
                .findFirst()
                .orElseThrow(
                        () ->
                                new AccountException(
                                        AccountException.ErrorCode.ACCOUNT_NOT_FOUND,
                                        "Account not found."));

        SpendingAnalysis analysis =
                spendingAnalysisQuery
                        .findAnalysis(accountId, analysisMonth)
                        .orElseThrow(
                                () ->
                                        new AccountException(
                                                AccountException.ErrorCode
                                                        .SPENDING_ANALYSIS_NOT_FOUND,
                                                "Spending analysis not found."));

        return new SpendingAnalysisResult(
                analysis.getAnalysisMonth(),
                analysis.getTotalAmount(),
                analysis.getTransactionCount(),
                analysis.getAverageAmount(),
                analysis.getChangeFromPreviousMonth(),
                analysis.getTrend().name(),
                analysis.getCreatedAt());
    }
}

File: implementations/kotlin-springboot/examples/src/main/kotlin/com/example/accountservice/account/application/query/GetSpendingAnalysisService.kt
package com.example.accountservice.account.application.query

import com.example.accountservice.account.domain.AccountFindQuery
import com.example.accountservice.account.domain.AccountNotFoundException
import com.example.accountservice.account.domain.SpendingAnalysisNotFoundException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate

/**
 * Serves the read-model row `AnalyzeMonthlySpendingService` (the ETL) already precomputed — this
 * Service does no aggregation of its own, it only looks the row up. Account (like Refund) has an
 * ownerId reachable in a single lookup, so ownership is verified via `AccountQuery.findAccounts`
 * first, the same pattern `GetAccountService`/`GetTransactionsService` use.
 */
@Service
@Transactional(readOnly = true)
class GetSpendingAnalysisService(
    private val accountQuery: AccountQuery,
    private val spendingAnalysisQuery: SpendingAnalysisQuery,
) {
    fun getSpendingAnalysis(
        accountId: String,
        requesterId: String,
        month: LocalDate,
    ): GetSpendingAnalysisResult {
        val (accounts, _) =
            accountQuery.findAccounts(
                AccountFindQuery(page = 0, take = 1, accountId = accountId, ownerId = requesterId),
            )
        accounts.firstOrNull() ?: throw AccountNotFoundException(accountId)

        val analysisMonth = "%04d-%02d".format(month.year, month.monthValue)
        val analyses = spendingAnalysisQuery.findAnalyses(SpendingAnalysisFindQuery(accountId, analysisMonth))
        val analysis = analyses.firstOrNull() ?: throw SpendingAnalysisNotFoundException(accountId, analysisMonth)

        return GetSpendingAnalysisResult(
            analysisMonth = analysis.analysisMonth,
            totalAmount = analysis.totalAmount,
            transactionCount = analysis.transactionCount,
            averageAmount = analysis.averageAmount,
            changeFromPreviousMonth = analysis.changeFromPreviousMonth,
            trend = analysis.trend.name,
            createdAt = analysis.createdAt,
        )
    }
}
